GH-1443: Fix default option when `--all-or-nothing` option is used as intended

agustingomes opened the pull request:
<!-- Fill in the relevant information below to help triage your pull request. -->

|      Q       |   A
|------------- | -----------
| Type         | bug
| BC Break     | no
| Fixed issues | https://github.com/doctrine/migrations/issues/1443
#### Summary

After a review of the issue reported (linked above), it seems the default behavior of specifying the option is opposite of what was intended, and also what is currently documented. This PR applies the correct default when the user specifies the `--all-or-nothing` option in the command line.

### Background:

Previous PR's and their linked issues contain more context regarding each iteration so far. They are presented below in the chronological order in which they were merged:
- https://github.com/doctrine/migrations/pull/1296
- https://github.com/doctrine/migrations/pull/1308
- https://github.com/doctrine/migrations/pull/1305
- https://github.com/doctrine/migrations/pull/1381

agustingomes committed:
Ad-Hoc: Add extra tests for configuration defaults

These extra tests may help in the future to catch unintended behavior changes for the default of these 2 specific settings.

agustingomes commented on lib/Doctrine/Migrations/Tools/Console/ConsoleInputMigratorConfigurationFactory.php:
```diff
@@ -45,15 +45,14 @@ private function determineAllOrNothingValueFrom(InputInterface $input): bool|nul
                     Problem: Passing values is deprecated
-                    Solution: If you need to disable the behavior, omit the option,
-                    otherwise, pass the option without a value
+                    Solution: how we should deal with this option being `true` in config but user want to disable it in command line for 4.x?
```
Added intentionally this line, since I'm wondering how to handle this.

Also as an extra thought, since we've added more coverage to how the option is determined, I wonder if it still makes sense to deprecate this ability, or if we still want to go ahead with disallowing values to be passed for this option.

MatTheCat replied:
What about a new `--no-all-or-nothing` option?

greg0ire replied:
> Also as an extra thought, since we've added more coverage to how the option is determined, I wonder if it still makes sense to deprecate this ability, or if we still want to go ahead with disallowing values to be passed for this option.

I don't think there's much value in having several ways of doing the same thing. It just adds complexity.

> What about a new --no-all-or-nothing option?

Might be good, and is easy to implement with https://symfony.com/blog/new-in-symfony-5-3-negatable-command-options

agustingomes replied:
@greg0ire that's a good idea, but it seems for our particular use case it won't work. looking at the docs:

> Negatable options are only available for options that don't allow passing any value to them (their previous type should be InputOption::VALUE_NONE).

For the remainder of the 3.x lifetime, it is better the users continue passing the optional value as it's currently allowed, and we introduce the negatable option in 4.x.

greg0ire replied:
Maybe we can use a negatable option in 4.x, and a full-fledged option in 3.x, literally named `no-all-or-nothing` to help with the transition?

agustingomes replied:
indeed, we can do that. my latest push should have a good starting point.

agustingomes commented:
@greg0ire The CI/CD failures do not seem directly related to my changes as far as I could see. Can you take a look if it's something I need to fix in my PR?

greg0ire commented:
You were targeting a branch that is no longer maintained. Please rebase.

greg0ire commented:
Still, the errors occur on 3.8.x, I'll try to address them.

agustingomes commented:
@greg0ire rebasing the branch and the extra commit fixed the CI issues.

greg0ire commented:
Sorry, I forgot to mention I made an attempt to address the SA issues in #1445

agustingomes committed:
GH-1443: fix `--all-or-nothing` default behavior

As result of some past changes to improve the way the option was inferred, several attempts to build a mechanism to handle the absence of value of this option were built via the following commits:

4da00c540672d687b8efa7b5f0c2b50107845379
533595134247c0c9f2e16c9a17fd1776a30303a4
5038099da818718703586804cc5bf890aec9086f
799f16d458a93aaadaec29012d8b8b425131fa1d
f726a5fdfeb9f61db594b436b62dbeeb5e925308

This commit leverages all the iterative improvements introduced along the way, and setting the correct value when the option is correctly specified.

This should bring the behavior inline with what is currently documented, while additionally adding a documentation deprecation informing users that passing a value to that option won't be allowed in 4.x

As an extra addition, this commit also introduces the negated option `--no-all-or-nothing`. The intent is to leverage the native Symfony >= 5.3 [Negatable command options](https://symfony.com/blog/new-in-symfony-5-3-negatable-command-options) in the future once the optional value from `--all-or-nothing` is completely removed.

agustingomes commented on src/Tools/Console/ConsoleInputMigratorConfigurationFactory.php:
```diff
@@ -31,30 +31,50 @@ public function getMigratorConfiguration(InputInterface $input): MigratorConfigu
+            $enableAllOrNothingOption = $input->getOption('all-or-nothing');
+        }
+
+        if ($enableAllOrNothingOption !== self::ABSENT_CONFIG_VALUE && $disableAllOrNothingOption === true) {
```
I will improve this logic, as it currently won't throw if we pass a value to `--all-or-nothing` option.

Edit: it was already correct. my bad.

greg0ire commented on docs/en/reference/configuration.rst:
```diff
@@ -243,18 +243,23 @@ Setting ``transactional`` to ``false`` will disable that.
 ~~~~~~~~~~~~~~~~~~~~~
 
-You can also set this option from the command line with the ``migrate`` command and the ``--all-or-nothing`` option:
+To enable all the migrations to complete together, you can also set this option from the command line with
```
I'm not sure people will understand what is meant by "completing together". I think we might not want to shy away from being technical here, talking about transactions should make clarify things IMO.

agustingomes replied:
Pushed a new phrasing, let me know if it's ok the way is written now.

greg0ire commented:
The SA issues are fixed, you can now rebase and drop your second commit.

agustingomes commented:
@greg0ire with the release of PHPStan 1.12 today a new error surfaced. How do you want to proceed?

SenseException commented:
> with the release of PHPStan 1.12 today a new error surfaced. How do you want to proceed?

It affects code that you haven't touched. It needs to be fixed, but not by this PR. See #1453.

greg0ire commented:
@agustingomes you may rebase :)
